File: prediction/src/prediction_validation.py
def is_valid_prediction(prediction: dict) -> bool:
    """Returns True when a prediction contains all required fields and valid data."""

    # Use camelCase for dictionary keys because predictions are shared logical objects
    # exchanged across multiple languages and should follow a consistent data schema.
    method = prediction.get("method")
    predicted_winner = prediction.get("predictedWinner")
    confidence_score = prediction.get("confidenceScore")

    # strip() ensures that all whitespace-only strings are rejected.
    if not isinstance(method, str) or not method.strip():
        return False

    if not isinstance(predicted_winner, str) or not predicted_winner.strip():
        return False

    # In Python, bool is a subclass of int, so boolean values must be rejected separately.
    if isinstance(confidence_score, bool) or not isinstance(
        confidence_score, (int, float)
    ):
        return False

    return 0.0 <= confidence_score <= 1.0

File: prediction/src/predictionValidation.js
export function isValidPrediction(prediction) {
  const method = prediction.method;
  const predictedWinner = prediction.predictedWinner;
  const confidenceScore = prediction.confidenceScore;

  // Reject missing, non-string, and whitespace-only values.
  if (typeof method !== "string" || !method.trim()) {
    return false;
  }

  if (typeof predictedWinner !== "string" || !predictedWinner.trim()) {
    return false;
  }

  if (typeof confidenceScore !== "number" || !Number.isFinite(confidenceScore)) {
    return false;
  }

  return confidenceScore >= 0.0 && confidenceScore <= 1.0;
}
